pageserver/src/tenant/layer_map.rs

```diff
@@ -137,6 +138,21 @@ where
         self.layer_map.remove_historic_noflush(layer)
     }
 
+    /// Replaces existing layer iff it is the `expected`.
+    ///
+    /// If the expected layer has been removed it will not be inserted by this function.
+    ///
+    /// Returns `true` if the layer map was changed, `false` otherwise. Errors are returned for
+    /// precondition failures, such as trying to replace two different kind of layers with each
+    /// other, and no modification is done in the case of precondition failure.
+    pub fn replace_historic(
```

Comment:
```suggestion
    ///
    /// TODO replacement can be done without buffering and rebuilding layer map updates.
    ///      One way to do that is to add a layer of indirection for returned values, so
    ///      that we can replace values only by updating a hashmap.
    pub fn replace_historic(
```

Worth commenting, in case this becomes an issue and someone else trips on it.